### PR TITLE
Chrome Custom Tabs

### LINKED ISSUE #93: Use Chrome Custom Tabs instead of WebView
Infinitely better!

https://developer.chrome.com/multidevice/android/customtabs

### PR DESCRIPTION
**Pull request by wiseodd**
Play around with the config

@timotiusnc review!

Closes #93

### CODE REVIEW

File: app/src/main/java/io/pumpkinz/pumpkinreader/util/ActionUtil.java [hunk 1]
```diff
@@ -73,20 +99,9 @@ public static void share(Context ctx, News news) {
 
         shareIntent.putExtra(Intent.EXTRA_TEXT, text);
         shareIntent.setType(Constants.MIME_TEXT_PLAIN);
+        shareIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/util/ActionUtil.java]: What's the usage of this additional flag?

File: app/src/main/java/io/pumpkinz/pumpkinreader/etc/Constants.java [hunk 2]
```diff
@@ -44,4 +45,9 @@
     public static final int TAB_WEBVIEW = 0;
     public static final int TAB_COMMENTVIEW = 1;
 
+    public static final String EXTRA_CUSTOM_TABS_SESSION = "android.support.customtabs.extra.SESSION";
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/Constants.java]: Are all these EXTRA_CUSTOM_TABS_\* unused?

**Comment by wiseodd** [comment]
Whoop yeah, I forgot to cleanup

File: app/src/main/java/io/pumpkinz/pumpkinreader/etc/PumpkinCustomTab.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    private CustomTabsIntent buildCustomTabsIntent() {
+        CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
+        builder.setShowTitle(true);
+        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(activity);
+
+        int color;
+
+        if (pref.getBoolean(Constants.CONFIG_DARK_THEME, false)) {
+            color = activity.getResources().getColor(R.color.primary_material_dark);
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/PumpkinCustomTab.java]: `getColor` is deprecated, bro.
http://stackoverflow.com/questions/31842983/getresources-getcolor-is-deprecated

File: app/src/main/java/io/pumpkinz/pumpkinreader/etc/PumpkinCustomTab.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        } else {
+            color = activity.getResources().getColor(R.color.pumpkin_primary);
+        }
+
+        builder.setToolbarColor(color);
+
+        String saveLabel = activity.getResources().getString(R.string.save);
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/PumpkinCustomTab.java]: This should be `shareLabel`, right? (and `shareIcon`)

File: app/src/main/java/io/pumpkinz/pumpkinreader/etc/ShareBroadcastReceiver.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    public void onReceive(Context context, Intent intent) {
+        News news = Parcels.unwrap(intent.getParcelableExtra(Constants.NEWS));
+
+        Intent chooser = Intent.createChooser(ActionUtil.getPumpkinShareIntent(context, news), "Share");
+        chooser.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+
+        context.startActivity(chooser);
+    }
+
+}
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/ShareBroadcastReceiver.java]: newline!

File: app/src/main/java/io/pumpkinz/pumpkinreader/etc/ShareBroadcastReceiver.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import io.pumpkinz.pumpkinreader.util.ActionUtil;
+
+
+public class ShareBroadcastReceiver extends BroadcastReceiver {
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        News news = Parcels.unwrap(intent.getParcelableExtra(Constants.NEWS));
+
+        Intent chooser = Intent.createChooser(ActionUtil.getPumpkinShareIntent(context, news), "Share");
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/ShareBroadcastReceiver.java]: Just call `ActionUtil.share`?

Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/etc/ShareBroadcastReceiver.java]: Oh, my bad, you need `FLAG_ACTIVITY_NEW_TASK`. Ok ok.

For `Share` string, how about using string from `strings.xml`?

**Comment by wiseodd** [comment]
Updated!

File: app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java [hunk 7]
```diff
@@ -154,8 +154,7 @@ public void goToNewsDetail(final News news) {
                 new Handler().postDelayed(new Runnable() {
                     @Override
                     public void run() {
-                        Uri uri = Uri.parse(news.getUrl());
-                        startActivity(new Intent(Intent.ACTION_VIEW, uri));
+                        ActionUtil.open(getActivity(), news);
                     }
                 }, 300);
             } else if (!shouldOpenLink) {
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java]: I just realized, this `else` is unused!

Comment (wiseodd) [inline app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java]: It's used when you don't enable "Open Link First" in the settings

File: app/src/main/java/io/pumpkinz/pumpkinreader/util/ActionUtil.java [hunk 8]
```diff
@@ -60,6 +70,22 @@ public void onClick(View view) {
     }
 
     public static void share(Context ctx, News news) {
```
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/util/ActionUtil.java:72]: The `ctx` variable is unused anymore.

File: app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java (hunk 7 above)
Comment (timotiusnc) [inline app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java]: but...but...line 150 has already initialized the `intent` variable

Comment (wiseodd) [inline app/src/main/java/io/pumpkinz/pumpkinreader/NewsListFragment.java]: Wuut you’re correct

**Comment by wiseodd** [comment]
OK all fixed

**Pull request merged**